### PR TITLE
Add force_result feature and automatic tests

### PR DESCRIPTION
**Pull request by perlpunk**
Issue: https://progress.opensuse.org/issues/102374

Tested with:

* Replace the `issues=` line with:
```
    issues='102374
foo bar auto_review:"Test died: Machine didn.t shut down":retry:force_result:failed'
```
* Call
```
    echo https://openqa.opensuse.org/tests/1990037 | ./openqa-label-known-issues
```

I also added automatic test, so far for label_on_issue and functions below.

I added [Test::More bash](https://github.com/ingydotnet/test-more-bash) as a subrepo, see README

### CODE REVIEW

File: openqa-label-known-issues
Comment (okurz) [inline openqa-label-known-issues:119]: Should we turn this around so that in case posting this fails at least is tried first?

**Review by okurz — CHANGES_REQUESTED** [review]
Looks nice. Can you also add a description in the README as documentation?

File: openqa-label-known-issues
Comment (perlpunk) [reply openqa-label-known-issues:119]: > Should we turn this around so that in case posting this fails at least is tried first?

I fail to parse this sentence...

Comment (okurz) [reply openqa-label-known-issues:119]: Nevermind. Just leave it like it is for now. We will see how it goes :)

**Review by kalikiana — APPROVED** [review]

**Comment by okurz** [comment]
Change is fine. Improve the commits and undraft for approval

**Comment by perlpunk** [comment]
@okurz  It's not yet finished. In the current state it's broken.

**Comment by perlpunk** [comment]
Now it's ready for review. I had to rearrange the functions a bit and put some of them into `_common` to be able to unit-test them, so unfortunately the change and the rearrangement is in one commit.
Added test-more-bash as a git-subrepo.

File: README.md
Comment (Martchus) [inline README.md:22]:
```suggestion
For this the subject line a ticket must include text following the format:
```

**Review by Martchus — COMMENTED** [review]
Looks generally good, only found a typo.

**Review by Martchus — APPROVED** [review]

File: Makefile
Comment (okurz) [inline Makefile:9]: `prove -r` should work

**Review by okurz — CHANGES_REQUESTED** [review]
I consider test-more-bash rather remote to our development (unlike what we do for os-autoinst-common) so I recommend against using subrepo to include that here. I would not even use git submodules here but simply rely on test-more-bash to be provided in a separate directory. Within the CI pipeline we can just clone it and within a local environment we can leave it to users to checkout but provide convenient setup tools, e.g. compare to https://github.com/os-autoinst/os-autoinst-distri-opensuse/blob/master/Makefile#L11

File: Makefile
Comment (kalikiana) [reply Makefile:9]: I'm guessing that's slower if it needs to search the whole tree?

File: Makefile
Comment (perlpunk) [reply Makefile]: done

Comment (perlpunk) [reply Makefile]: done

File: Makefile
Comment (perlpunk) [reply Makefile]: I made it `prove -r test/` now.

**Comment by perlpunk** [comment]
> I consider test-more-bash rather remote to our development

I don't see that as a reason against subrepo. The repo is not that big.
Opinions from anyone else?

**Comment by kalikiana** [comment]
> > I consider test-more-bash rather remote to our development
> 
> I don't see that as a reason against subrepo. The repo is not that big. Opinions from anyone else?

Before that I would ask why this isn't packaged. If we mainly want to make it *available* rather than develop and it's used like a *library* I would regard it as an external dependency first.

File: Makefile
Comment (okurz) [reply Makefile]: I would have not expected anything else because it's not automatically tested so will rot :) still serves as an example

**Comment by Martchus** [comment]
I approved the PR because I'm not strictly against using Git subrepo here. However, I would also prefer to keep things separate here for reasons mentioned by @okurz . In addition, the subrepo approach is clearly *not* scaling as it is merely copying the contents after all. Having to do a separate checkout (e.g. just next to the repo) is not the worst if it is documented. It can even be simpler for the user because there are no new commands to learn about.

**Comment by kalikiana** [comment]
> That is not really helpful though. It isn't packaged and doing that is certainly more work than subrepo/submodule/clone

If you already discussed this, you could have mentioned it 😉️

No objections from me then. The subrepo ensures that we consistently use the same version, especially if it isn't packaged.

**Review by kalikiana — DISMISSED** [review]
> That is not really helpful though. It isn't packaged and doing that is certainly more work than subrepo/submodule/clone

If you already discussed this, you could have mentioned it 😉️

No objections from me then. The subrepo ensures that we consistently use the same version, especially if it isn't packaged.

**Comment by perlpunk** [comment]
I will remove the subrepo.

Just commenting on the arguments:

> [A separate checkout] can even be simpler for the user because there are no new commands to learn about.

There is nothing to learn for the user. It's a `git subrepo clone` once, and no user has to care about it ever. For the user, it's the **easiest** solution.

Maybe you mean the maintainers.
Having a problem with learning new commands would speak against every new technology.

**Comment by perlpunk** [comment]
I removed it. It will now be cloned when running `make test-unit`.

(If test-more-bash is ever updated and we want the new version, we need to make sure that the version is correct if it has already been cloned before.)

**Review by okurz — APPROVED** [review]

File: README.md
Comment (Martchus) [inline README.md:142]:
```suggestion
It will be automatically cloned.
```

**Review by Martchus — APPROVED** [review]

File: README.md
Comment (perlpunk) [reply README.md]: done

File: .github/workflows/ci.yml
Comment (okurz) [inline .github/workflows/ci.yml:14]: As you add a dedicated section here please also extend .mergify.yml in https://github.com/os-autoinst/scripts/blob/master/.mergify.yml#L11 and accordingly https://github.com/os-autoinst/scripts/blob/master/.mergify.yml#L20 with

```patch
       - "status-success~=test"
+      - "status-success~=style"
```

otherwise mergify would merge if tests pass but the style checks which you extended into a separate section could still fail. As an alternative of course you could keep them combined under simply "test". By the way, I would also prefer to not need to explicitly spell out each kind of test in .mergify as this is error prone as soon as we introduce another test job. In os-autoinst and openQA we could go with the alternative to look for the codecov report because only if *all* tests pass then also codecov is called. Maybe eventually we can reach something similar here as well with a fan-in job that combines all results into a final status check that we can check for in mergify

**Review by okurz — CHANGES_REQUESTED** [review]
> > That is not really helpful though. It isn't packaged and doing that is certainly more work than subrepo/submodule/clone
> 
> If you already discussed this, you could have mentioned it wink

Well, we haven't discussed this in before just that I think asking for an external dependency to be packaged first is IMHO too much effort to get this PR merged.

@kalikiana your review is pending, would you like to approve this as well?

File: .github/workflows/ci.yml
Comment (perlpunk) [reply .github/workflows/ci.yml:14]: Ok. I really like that the different tests are displayed on their own, so it's easy to see where the problem is.

Adding a test that succeeds when all others succeeded should be possible. But maybe for another PR in the future.

**Review by okurz — APPROVED** [review]

**Review by okurz — CHANGES_REQUESTED** [review]
argh, wait. You need to touch *both* L11 and L22 in .mergify.yml

File: .github/workflows/ci.yml
Comment (okurz) [reply .github/workflows/ci.yml:14]: please also add for https://github.com/os-autoinst/scripts/blob/master/.mergify.yml#L20

Comment (perlpunk) [reply .github/workflows/ci.yml:14]: oh sorry. done

**Review by okurz — APPROVED** [review]

**Comment by okurz** [comment]
With the changes to mergify config we need to merge manually

**Pull request merged**
